src/org/sosy_lab/java_smt/test/SolverConcurrencyTest.java

```diff
@@ -0,0 +1,437 @@
+
+  private static List<Throwable> exceptionsList;
+
+  @Parameters(name = "{0}")
+  public static Object[] getAllSolvers() {
+    return Solvers.values();
+  }
+
+  @Parameter(0)
+  public Solvers solver;
+
+
+  protected Solvers solverToUse() {
+    return solver;
+  }
+
+  @BeforeClass
+  public static void createExceptionsList() {
+    exceptionsList = Collections.synchronizedList(new ArrayList<Throwable>());
+  }
+
+  @After
+  public void resetExceptionsList() {
+    exceptionsList.clear();
+  }
+
+  @Test
+  public void testIntConcurrencyWithConcurrentContext() {
+    // Test concurrency of integers (while every thread creates its unique context on its own
+    // concurrently)
+    assume().withMessage("Solver %s does not support the theory of bitvectors", solverToUse())
+        .that(solverToUse())
+        .isNotEqualTo(Solvers.BOOLECTOR);
+    List<Runnable> runnableList = new ArrayList<>();
+    for (int i = 0; i < NUMBER_OF_THREADS; i++) {
+      Runnable test = new Runnable() {
+        @SuppressWarnings("resource")
+        @Override
+        public void run() {
+          try {
+            SolverContext context = initSolver();
+            intConcurrencyTest(context);
+            closeSolver(context);
+          } catch (Throwable e) {
+            exceptionsList.add(e);
+          }
+        }
+      };
+      runnableList.add(test);
+    }
+    assertConcurrency(runnableList, "testBasicIntConcurrency");
+  }
+
+  @Test
+  public void testBvConcurrencyWithConcurrentContext() {
+    // Test concurrency of bitvectors (while every thread creates its unique context on its own
+    // concurrently)
+    assume().withMessage("Solver %s does not support the theory of bitvectors", solverToUse())
+        .that(solverToUse())
+        .isNotEqualTo(Solvers.SMTINTERPOL);
+    List<Runnable> runnableList = new ArrayList<>();
+    for (int i = 0; i < NUMBER_OF_THREADS; i++) {
+      Runnable test = new Runnable() {
+        @SuppressWarnings("resource")
+        @Override
+        public void run() {
+          try {
+            SolverContext context = initSolver();
+            bvConcurrencyTest(context);
+            closeSolver(context);
+          } catch (Throwable e) {
+            exceptionsList.add(e);
+          }
+        }
+      };
+      runnableList.add(test);
+    }
+    assertConcurrency(runnableList, "testBasicIntConcurrency");
+  }
+
+  @SuppressWarnings("resource")
+  @Test
+  public void testIntConcurrencyWithoutConcurrentContext() throws InvalidConfigurationException {
+    // Test concurrency with already present and unique context per thread
+    assume().withMessage("Solver %s does not support the theory of bitvectors", solverToUse())
+        .that(solverToUse())
+        .isNotEqualTo(Solvers.BOOLECTOR);
+    List<Runnable> runnableList = new ArrayList<>();
+    ConcurrentLinkedQueue<SolverContext> contextList = new ConcurrentLinkedQueue<>();
+    for (int i = 0; i < NUMBER_OF_THREADS; i++) {
```

Comment:
Splitting the loop into two (initSolvers and getRunnables) would make it clearer that initSolvers is sequential and the other is not.

Reply:
Done